[experiment] Add `-Z no-link` flag

0dvictor opened the pull request:
Adds a compiler option to allow rustc compile a crate without linking.
With this flag, `rustc` serializes codegen_results into a `.rlink` file.

Part of Issue #64191

rust-highfive commented:
r? @matthewjasper

(rust_highfive has picked a reviewer for you, use r? to override)

rust-highfive commented:
The job `x86_64-gnu-llvm-7` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=15600&view=logs&jobId=0230d20d-e0df-5b62-a6ec-fad0f9c3c3d5) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/15600/logs/65), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/15600/logs/65)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-12-10T12:35:15.6609777Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-12-10T12:35:15.6817518Z ##[command]git config gc.auto 0
2019-12-10T12:35:16.2452065Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-12-10T12:35:16.2457204Z ##[command]git config --get-all http.proxy
2019-12-10T12:35:16.2463412Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/67195/merge:refs/remotes/pull/67195/merge
---
2019-12-10T12:41:09.8847172Z    Compiling serde_json v1.0.40
2019-12-10T12:41:11.6221690Z    Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
2019-12-10T12:41:22.6207840Z     Finished release [optimized] target(s) in 1m 28s
2019-12-10T12:41:22.6322224Z tidy check
2019-12-10T12:41:23.4268625Z tidy error: /checkout/src/librustc_session/config.rs: too many lines (3015) (add `// ignore-tidy-filelength` to the file to suppress this error)
2019-12-10T12:41:25.3800276Z some tidy checks failed
2019-12-10T12:41:25.3803694Z Found 485 error codes
2019-12-10T12:41:25.3804509Z Found 0 error codes with no tests
2019-12-10T12:41:25.3804796Z Done!
2019-12-10T12:41:25.3804796Z Done!
2019-12-10T12:41:25.3805117Z 
2019-12-10T12:41:25.3805345Z 
2019-12-10T12:41:25.3806701Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor"
2019-12-10T12:41:25.3809219Z 
2019-12-10T12:41:25.3809446Z 
2019-12-10T12:41:25.3809791Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
2019-12-10T12:41:25.3810052Z Build completed unsuccessfully in 0:01:32
2019-12-10T12:41:25.3810052Z Build completed unsuccessfully in 0:01:32
2019-12-10T12:41:25.3864524Z == clock drift check ==
2019-12-10T12:41:25.3872832Z   local time: Tue Dec 10 12:41:25 UTC 2019
2019-12-10T12:41:25.5375810Z   network time: Tue, 10 Dec 2019 12:41:25 GMT
2019-12-10T12:41:25.5378009Z == end clock drift check ==
2019-12-10T12:41:26.8769521Z 
2019-12-10T12:41:26.8887665Z ##[error]Bash exited with code '1'.
2019-12-10T12:41:26.8924335Z ##[section]Starting: Checkout
2019-12-10T12:41:26.8926199Z ==============================================================================
2019-12-10T12:41:26.8926260Z Task         : Get sources
2019-12-10T12:41:26.8926311Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bjorn3 commented on src/librustc_interface/passes.rs:
```diff
@@ -933,6 +933,10 @@ fn encode_and_write_metadata(
 
     let metadata = match metadata_kind {
+
+        // Metadata is always required for two-staged compile and link.
```
The only necessary data is the data passes to https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_utils/codegen_backend/trait.CodegenBackend.html#tymethod.join_codegen_and_link:

* Session
* OutputFilenames
* [CodegenResults](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_ssa/struct.CodegenResults.html) (Actually OngoingCodegen is passed, but when joining the codegen, CodegenResults is produced)

DepGraph is not necessary (it is used for storing the location of incremental compilation artifacts, https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_codegen_llvm/lib.rs.html#307 but after joining codegen, the no-link invocation of rustc can do this itself)

0dvictor replied:
Yes, metadata is for reconstructing CrateInfo (part of CodegenResult), as it need information about dependent libraries (both rlib and native ones).

bjorn3 replied:
It should be possible to just derive `Serialize` for `CrateInfo` and put it in a `.rlink` file together with the other necessary data.

0dvictor replied:
Good idea, let me look into it.

0dvictor replied:
I found `CrateInfo` + "other necessary data" is indeed `CodegenResults`. So I updated my the PR to serialize `CodegenResults` into a `.rlink` file.

bjorn3 commented on src/librustc_codegen_llvm/lib.rs:
```diff
@@ -308,7 +308,8 @@ impl CodegenBackend for LlvmCodegenBackend {
+        if sess.opts.debugging_opts.no_link ||
+           !sess.opts.output_types.keys().any(|&i| i == OutputType::Exe ||
                                                    i == OutputType::Metadata) {
             return Ok(());
```
I noticed that `finalize_session_directory` is never called in this case, which means that incremental compilation won't work.

tmandry replied:
Should it have been called for `Exe` and `Metadata` output types, too?

bjorn3 replied:
The current condition says return if neither exe, nor metadata is emitted.

bjorn3 replied:
I just realized that `codegen_backend.join_codegen_and_link` needs to be called to actually emit all object files. Maybe split rename it to `join_codegen` and return `CodegenResults` from it, leaving the linker part to `rustc_driver`?

0dvictor replied:
Good suggestion. In fact, I think it is a good refactoring even without this `-Z no-link` change.

tmandry commented:
r? @tmandry
cc @alexcrichton

Centril commented on src/librustc_codegen_ssa/back/write.rs:
```diff
@@ -538,6 +560,32 @@ fn produce_final_output_artifacts(sess: &Session,
+                    remove(sess, &path);
+                }
+            }
+            if let Some(ref module) = compiled_modules.metadata_module {
```
The structure of this code looks similar to the bit above (and same below); it would be good to avoid copy paste programming.

0dvictor replied:
Totally agreed, but not doing so would require refactoring existing code a bit. They are actually copied from existing code.
I'd like to avoid refactoring and new feature in the same commit or pull request. Let me create a separate PR for the refactoring and then rebase this PR.

Centril replied:
Certainly. I only care that the work to refactor eventually gets done. Feel free to structure your refactorings as you see fit over PRs / commits. :)

0dvictor replied:
I've updated the changes based on reviewers' feedbacks, and this file won't be touched anymore.

tmandry commented on src/librustc_codegen_ssa/back/write.rs:
```diff
@@ -390,10 +390,26 @@ pub fn start_async_codegen<B: ExtraBackendMethods>(
+            OutputType::Bitcode => {
+                modules_config.emit_bc = true;
+                if sess.opts.debugging_opts.no_link {
+                    metadata_config.emit_bc = true;
```
I may be missing something here, but I don't really see why metadata codegen involves bitcode, LLVM IR, etc. This should only control the output of the `rmeta` file read by rustc if I'm not mistaken.

bjorn3 replied:
I believe `-Zno-link` should still compile everything, only leaving the linking to a next rustc invocation.

0dvictor replied:
This metadata_config isn't for the `rmeta` file used later for linking. It is to [generate a metadata module linking into a `Dylib` or `ProcMacro`](https://github.com/rust-lang/rust/blob/f284f8b4be3a899bf2ecc03e2a1589f486b62a9f/src/librustc_interface/passes.rs#L911-L969). Bitcode/LLVM IR/etc. won't be generated for any other crate types.

tmandry commented:
It.. looks like this change hit a file line length tidy limit (3000 lines) in `src/librustc_session/config.rs`.

This seems arbitrary, and it's a bit unfair for a new contributor to have to deal with this. Does anyone who's contributed to this file have suggestions on how to break it up? Or should we override it for now?

Centril commented:
@tmandry The limit is arbitrary, but that's the fundamental nature of limits. :) I believe these tidy checks are necessary for code quality to push us towards reasonably sized files (not like `parser.rs` was before...).

I would suggest one of the following strategies:

- Refactor the file before landing this PR and then rebase this one. (This is probably not super fun but perhaps @0dvictor wants to tackle it?)

- Refactor the file after landing this PR and leave an override for now; file an issue to split it up and assign yourself to do it. It shouldn't take a long time to split up a file and reviewing it is just a matter of using `diff` to check that nothing really changed. Moving out `fn build_session_options` and friends seem like a prime candidate for splitting up. I'd also consider moving the data types + some impls out to its own module.

0dvictor commented:
I'm happy to refactor `config.rs` then rebase this PR. Briefly going through this file, I found it now contains two kinds of logic: 1. definition of CLI arguments, and 2. logics to parse and validate CLI arguments. So I'm thinking of moving following code out to a new file named `options.rs`:
* https://github.com/rust-lang/rust/blob/f284f8b4be3a899bf2ecc03e2a1589f486b62a9f/src/librustc_session/config.rs#L425-L508
* https://github.com/rust-lang/rust/blob/f284f8b4be3a899bf2ecc03e2a1589f486b62a9f/src/librustc_session/config.rs#L1235-L1573

On top of that, as @Centril suggested, I also would like to move `fn build_session_options` and friends into a new file named `config_builder.rs`.

What do you guys think, @tmandry and @Centril ?

Centril commented:
> What do you guys think, @tmandry and @Centril ?

Sounds like a sensible general direction. 👍

tmandry commented:
I don't personally love the idea of separating the primary data structure definition of `Options` from the code that builds it. I am fine with separating out the macros and validator helper functions. So I guess I would personally put both the things you're talking about splitting into the same file.

That said, just do what seems most reasonable for now and we can see how it looks in review.

alexcrichton commented:
It may be worth pointing out that simply the presence of `--emit obj` will force codegen units to 1, so if `-Z no-link` is tweaking the output types of the compiler it may have the side effect of pretty radically changing compilation by reducing the number of codegen units to 1 in all cases.

0dvictor commented:
[`// ignore-tidy-filelength`](https://github.com/rust-lang/rust/blob/12307b3b08edee543a78fb9d4a837fbd6d6ac0fa/src/librustc_session/config.rs#L1) was introduced by #67074 to `config.rs`, do we still want to refactor this file? @Centril @tmandry

0dvictor commented:
> It may be worth pointing out that simply the presence of --emit obj will force codegen units to 1, so if -Z no-link is tweaking the output types of the compiler it may have the side effect of pretty radically changing compilation by reducing the number of codegen units to 1 in all cases.

Thank you for bring this up. Yes, at the moment, `-Z no-link` would only allow 1 codegen units. I am working on enhancing it to support multiple CGUs. It'll basically make `--emit obj` to support multiple CGUs.

Centril commented:
> `config.rs`, do we still want to refactor this file? @Centril @tmandry

@0dvictor Yes, definitely, and I think that:

> I don't personally love the idea of separating the primary data structure definition of `Options` from the code that builds it. I am fine with separating out the macros and validator helper functions. So I guess I would personally put both the things you're talking about splitting into the same file.

...is not something I agree with. :) In particular, it makes a whole lot of sense to separate data definitions and logic so that things are more pluggable and it's also better in general for incremental. This is precisely the thing we've been trying to do with splitting libsyntax from the parser. So I want to encourage your refactoring proposal.

alexcrichton commented:
@0dvictor 

> It'll basically make `--emit obj` to support multiple CGUs.

As a heads up that's unfortunately a breaking change since `--emit obj` only emits one object file today, so there likely needs to be a different route to implement this. I think that we'll benefit from a more "first class" implementation as well rather than trying to patch together a few other options to make this work.

bjorn3 commented:
Maybe have something like `--emit delayed-link` (not `--emit obj`) to emit all object files necessary for linking combined with a `.rlink` file and then something like `rustc crate.rlink` to perform the linking.

0dvictor commented:
> As a heads up that's unfortunately a breaking change since --emit obj only emits one object file today, so there likely needs to be a different route to implement this.

Good point. Let me think about it and update the changes.

BTW, unrelated to `-Z no-link` but out of curiosity: if `--emit obj` should always generate one file, why do we have to disable parallel compilation? Technically we could still compile in parallel with multiple CGUs, but at the end, we can combine all CGUs with `llvm::Linker` so that we still get one obj file.

0dvictor commented:
> Maybe have something like --emit delayed-link (not --emit obj) to emit all object files necessary for linking combined with a .rlink file and then something like rustc crate.rlink to perform the linking.

Good suggestion, and it should work well with your previous suggestion on serializing `CrateInfo`. Working on it.

alexcrichton commented:
@0dvictor that's possible yeah! It's not as parallel since codegen is still serial (one big object file), but it's probably best to leave all this to a separate issue.

bors commented:
:umbrella: The latest upstream changes (presumably #67449) made this pull request unmergeable. Please resolve the merge conflicts.

0dvictor commented on src/librustc/hir/def_id.rs:
```diff
@@ -84,8 +85,16 @@ impl fmt::Display for CrateNum {
-impl rustc_serialize::UseSpecializedDecodable for CrateNum {}
+impl rustc_serialize::UseSpecializedEncodable for CrateNum {
+    fn default_encode<E: Encoder>(&self, e: &mut E) -> Result<(), E::Error> {
+        e.emit_u32(self.as_u32())
```
Thank you for bringing this up. I was actually struggled a lot trying to emit the correct data. I believe we can safely simply emit the crate number here.
As you pointed out, the crate numbers is only meaningful within the same compilation session; therefore, one must have the session info when using crate numbers. For example, one might need include a crate number -> `Svh` mapping table in the serialized data, or some other ways to correctly map crate numbers to the actual crates.

In this PR, crate numbers are used within one `CodegenResults`, and `CodegenResults` only belongs to one specific session; therefore, crate numbers are unique. Then we can use `CodegenResults.crate_info.used_crate_source` to find the crate from a crate number.

bjorn3 replied:
You also replaced the `Encodable` impl of `CrateNum` for other usages. Maybe make a new `LinkingCrateNum` newtype for `u32` and store that in `crate_info` instead?

0dvictor replied:
I don't think this `Encodable` impl is currently used since the [default impl](https://github.com/rust-lang/rust/blob/9ff30a7810c586819a78188c173a7b74adbb9730/src/libserialize/serialize.rs#L889-L894) just returns an error.
However, I like your idea of making a new `LinkingCrateNum` as it conceptually decouples `CrateInfo` from a compilation session. I'll update accordingly.

0dvictor replied:
I created #67516 for decoupling `CrateNum` from the compilation session. @bjorn3 could you take a look?

0dvictor commented on src/librustc/hir/def_id.rs:
```diff
@@ -152,8 +161,21 @@ impl DefId {
+    fn default_encode<S: Encoder>(&self, s: &mut S) -> Result<(), S::Error> {
+        let krate = u64::from(self.krate.as_u32());
+        let index = u64::from(self.index.as_u32());
+        s.emit_u64((krate<<32) | index)
```
Similar to `CrateNum`, one should include a `DefId` to `DefPath` mapping in the serialized data. However, in this PR, `DefId` isn't used by the linker so that the mapping can be safely omitted.

bors commented:
:umbrella: The latest upstream changes (presumably #67540) made this pull request unmergeable. Please resolve the merge conflicts.

JohnCSimon commented:
Ping from triage: @0dvictor can you please address the merge conflicts? Thanks.

0dvictor commented:
> Ping from triage: @0dvictor can you please address the merge conflicts? Thanks.

Thank you for reminding. I have rebased and resolved the conflicts; initially, I was plan to wait for the review of #67516, as it is to resolve concerns of this PR.

rust-highfive commented:
The job `x86_64-gnu-llvm-7` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=18831&view=logs&jobId=0230d20d-e0df-5b62-a6ec-fad0f9c3c3d5) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/18831/logs/66), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/18831/logs/66)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2020-01-19T12:10:44.8229490Z ========================== Starting Command Output ===========================
2020-01-19T12:10:44.8269317Z [command]/bin/bash --noprofile --norc /home/vsts/work/_temp/d96239a8-6e89-49d2-9276-aae83c16a230.sh
2020-01-19T12:10:44.8269535Z 
2020-01-19T12:10:44.8273180Z ##[section]Finishing: Disable git automatic line ending conversion
2020-01-19T12:10:44.8279510Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/67195/merge to s
2020-01-19T12:10:44.8281661Z Task         : Get sources
2020-01-19T12:10:44.8281694Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-01-19T12:10:44.8281773Z Version      : 1.0.0
2020-01-19T12:10:44.8281806Z Author       : Microsoft
---
2020-01-19T12:10:45.8802517Z ##[command]git remote add origin https://github.com/rust-lang/rust
2020-01-19T12:10:45.8822915Z ##[command]git config gc.auto 0
2020-01-19T12:10:45.8827461Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2020-01-19T12:10:45.8829833Z ##[command]git config --get-all http.proxy
2020-01-19T12:10:45.8841907Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/67195/merge:refs/remotes/pull/67195/merge
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

0dvictor commented:
Rebase again to fix the tidy error.

tmandry commented on src/librustc/middle/cstore.rs:
```diff
@@ -160,6 +160,7 @@ pub enum ExternCrateSource {
     Path,
 }
 
+#[derive(RustcEncodable, RustcDecodable)]
```
I'd prefer to save this in a binary file rather than escaped binary in a json string.

That said, I'm okay with doing this later, especially since this is an experimental feature.

0dvictor replied:
Totally agreed, a binary file is definitely the proper way. I chose JSON in this PR purely because I don't want the file format details hold the feature, especially the proposed approach has not been reviewed yet.

tmandry commented on src/librustc_hir/def_id.rs:
```diff
@@ -86,8 +87,16 @@ impl fmt::Display for CrateNum {
 
-impl rustc_serialize::UseSpecializedEncodable for CrateNum {}
-impl rustc_serialize::UseSpecializedDecodable for CrateNum {}
+impl rustc_serialize::UseSpecializedEncodable for CrateNum {
```
Since we decided to stick with this approach (https://github.com/rust-lang/rust/pull/67516#issuecomment-576695916), we may want to add a comment (here or on `CrateNum`) saying which conditions have to hold when using this.

0dvictor commented:
> what's the status of this? It doesn't look like you're changing the output types anymore. Does that mean multiple CGUs are supported now?

Correct, multiple CGUs are naturally supported; no extra work is needed.

> There was some discussion about `--emit delayed-link` but I'm not sure if it's needed. To my eyes everything looks good.

Yes, though I like the idea of `--emit delayed-link`, I don't think it is proper at current stage. As an experimental/unstable feature, a `-Z` option serves better.  I would like to discuss about it more when we decide to stabilize the feature.

bjorn3 commented:
It is possible to hide non `-Z` options behind `-Zunstable-options`.

tmandry commented on src/librustc_codegen_ssa/lib.rs:
```diff
@@ -117,6 +117,12 @@ bitflags::bitflags! {
 }
 
 /// Misc info we load from metadata to persist beyond the tcx.
+/// Note: though `CrateNum` is only meaningful within the same tcx, information within `CrateInfo`
```
```suggestion
///
/// Note: though `CrateNum` is only meaningful within the same tcx, information within `CrateInfo`
```

tmandry commented:
@bors r+

bors commented:
:pushpin: Commit be86fb3516135944944318b873e6f3cfe0a599b0 has been approved by `tmandry`

<!-- @bors r=tmandry be86fb3516135944944318b873e6f3cfe0a599b0 -->
<!-- homu: {"type":"Approved","sha":"be86fb3516135944944318b873e6f3cfe0a599b0","approver":"tmandry"} -->

bors commented:
:hourglass: Testing commit be86fb3516135944944318b873e6f3cfe0a599b0 with merge 089b0632c6ece90649d7b94596970f458cb9d292...
<!-- homu: {"type":"BuildStarted","head_sha":"be86fb3516135944944318b873e6f3cfe0a599b0","merge_sha":"089b0632c6ece90649d7b94596970f458cb9d292"} -->

tmandry commented:
@bors r+

bors commented:
:pushpin: Commit 6a6ebb4403683e1e12d3916cabc1a4898ce798cf has been approved by `tmandry`

<!-- @bors r=tmandry 6a6ebb4403683e1e12d3916cabc1a4898ce798cf -->
<!-- homu: {"type":"Approved","sha":"6a6ebb4403683e1e12d3916cabc1a4898ce798cf","approver":"tmandry"} -->

bors commented:
:hourglass: Testing commit 6a6ebb4403683e1e12d3916cabc1a4898ce798cf with merge 26fe0c8fa84a65deb9fe8ee6f2b928ed817c1d0b...
<!-- homu: {"type":"BuildStarted","head_sha":"6a6ebb4403683e1e12d3916cabc1a4898ce798cf","merge_sha":"26fe0c8fa84a65deb9fe8ee6f2b928ed817c1d0b"} -->

0dvictor commented:
Thank you all for reviewing and guiding me through the process. I just accepted the suggested changes, squashed all commits into one, and rebased to the latest. The PR is good to go.

tmandry commented:
@bors retry rolled up

rust-highfive commented:
Your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=19133&view=logs&jobId=1c64c3ff-4c37-56fe-a133-b407354bbfbf) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/19133/logs/130), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/19133/logs/130)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:hourglass: Testing commit 6a6ebb4403683e1e12d3916cabc1a4898ce798cf with merge e23dd6687f7677a715765bff1fe5e63b906cb96b...
<!-- homu: {"type":"BuildStarted","head_sha":"6a6ebb4403683e1e12d3916cabc1a4898ce798cf","merge_sha":"e23dd6687f7677a715765bff1fe5e63b906cb96b"} -->
